src/validation.cpp

```diff
@@ -1695,8 +1695,9 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
         return PackageMempoolAcceptResult(package_state_quit_early, {});
     }
 
-    // IsChildWithParents() guarantees the package is > 1 transactions.
-    assert(package.size() > 1);
```

Comment:
Won't we redundantly call `AcceptSubPackage` twice for a single transaction, if `individual_results_nonfinal.emplace(wtxid, single_res);` stuff is triggered?

Reply:
you're right, relatively harmless, but pushed a change which means it won't try package evaluation unless `txns_package_eval` is larger than size 1.

Reply:
It also causes double-validation for single transaction package that is missing inputs.

Could special case this but not sure I love making things even harder to understand:
```
diff --git a/src/validation.cpp b/src/validation.cpp
index 3e48335255..c89050e895 100644
--- a/src/validation.cpp
+++ b/src/validation.cpp
@@ -1790,10 +1790,16 @@ PackageMempoolAcceptResult MemPoolAccept::AcceptPackage(const Package& package,
                 // some of them may still be valid.
                 quit_early = true;
                 package_state_quit_early.Invalid(PackageValidationResult::PCKG_TX, "transaction failed");
                 individual_results_nonfinal.emplace(wtxid, single_res);
             } else {
+                // It's retryable in a package but this is a singleton so we should mark the package failed
+                if (package.size() == 1) {
+                    quit_early = true;
+                    package_state_quit_early.Invalid(PackageValidationResult::PCKG_TX, "transaction failed");
+                }
+
                 individual_results_nonfinal.emplace(wtxid, single_res);
                 txns_package_eval.push_back(tx);
             }
         }
     }
```

thoughts?

Reply:
I've failed to suggest a simple code-change to avoid double-validation of a single tx.

The code is really about multi-tx. Even if we add a bunch of conditions around `auto multi_submission_result`, stuff below (`individual_results_nonfinal`) makes you think about multi-tx again.

The least thing I'm worried about is double-validation on itself.
I'm worried about code comprehension complexity: reading multi-tx code against a single-tx scenario.

At this point I agree with Gloria to handle this case earlier. Or substantially moving code around inside this function to make reasoning about 1-conf more natural to it.

Reply:
The fix for double-validation seems good.
--------------
Do you mind sharing the code of your attempt (assuming it compiles :))? I wanna take a look, but would be useful for archival reasons anyway.

Reply:
> Do you mind sharing the code of your attempt (assuming it compiles :))? I wanna take a look, but would be useful for archival reasons anyway.

It compiled, but it won't pass tests because I would have had to heavily modify the BroadcastTransaction interface. Not sure I kept the branch sorry.

Reply:
So you'd rather not take `else if (package.size() == 1 || //` still? I think it's an improvement for code comprehension if we end up handling single-tx here.

Reply:
Expanding on this thought, I sketched a two-step improvement in handling single-package tx wrt code readability. Wanna [take a look](https://github.com/naumenkogs/bitcoin/commits/2024-10-submitpackage-singleton-v1/)?

I certainly might be missing something.

Reply:
Left a comment, I think my one-liner is the cleanest correct solution?

pushed it as a separate commit

Reply:
Have you left a comment on my branch? I can't see it for some reason.

Reply:
https://github.com/naumenkogs/bitcoin/commit/88c10c5789bfae1865d03966d1ce523e870ff0bc#diff-97c3a52bc5fad452d82670a7fd291800bae20c7bc35bb82686c2c0a4ea7b5b98R1811 weird comment UI, here is a direct link hopefully

Reply:
Weird UI indeed, i left a reply and I'll try to monitor that thread now.